### DISCUSSION TITLE
How to go about visually designing a significant new feature

### DISCUSSION THREAD

**Discussion by ebwinters** [bluesky-social/atproto#3288] [How to do it?]
I'm wondering if there is a process for making a significant change to the UI. In my case, adding a bookmarking feature

## Background
There have been discussions around bookmarks (#2000) being implemented at a protocol level rather than in Bluesky's lexicon. The community lexicon has implemented a bookmarking lexicon and currently I have a [PR open for the associated typescript client code](https://github.com/lexicon-community/js-lexicon-community-types/pull/2). Once that code is merged, I want to start working on a bookmarks implementation using this common lexicon inside of Bluesky. There is already a [PR open for adding the bookmarking icons](https://github.com/bluesky-social/social-app/pull/7231#pullrequestreview-2518555885), but I don't know if it's the right icon library.

## The problem
I have an idea of how I'd like to implement the feature at the API level, with something like
```ts
const record = {
    $type: 'community.lexicon.bookmarks.bookmark',
    subject: 'https://some-link-or-at-uri.xyz',
    tags: ['bluesky-bookmark', 'some_other_tag']
    createdAt: new Date().toISOString(),
}
await agent.com.atproto.repo.putRecord({
      repo: agent.assertDid,
      collection: 'community.lexicon.bookmarks.bookmark',
      rkey,
      record,
      validate: false,
})
```
But when it comes to design, I have my own ideas of how to show bookmarks in the Bluesky app view. There are a bunch of questions like
1. Should we show total bookmark count like likes count on a post?
2. Should we show bookmarks on a user's profile publicly or just for them to view
3. How should we organize bookmarks with tags or just show a list of all a user's bookmarks
4. Should we only show app.bsky related bookmarks?

I am happy to write a proposal and make some rough mockups, but I am wondering if there is a more defined process or document I should be looking at?

**Comment by SaySayTakamura** [comment]
If you allow me to answer Questions 1 and 3

1 -Showing the amount of posts Bookmarked is a good idea, i think formatting the AppBar Title would be a good way of implementing this, like: 

```
var title = $"Bookmarks - (Total: {totalPostsBookmarked})";
```

Following Twitter version of the Bookmarks view you could add a button to wipe out all the Bookmarked items!

3 - Personally, i'd go with Folders instead of Tags, Folders could allow the user to customize it with, Color, Name and maybe a description.

You could use Tags as Folders as well (but only having a name (the tag itself), no customization at all), that it's up to you.

I even made an issue on the [App Repo](https://github.com/bluesky-social/social-app/issues/1527#issue-1913214940) about this.

I think i can answer Question 2 as well, i believe showing the Bookmarks publicly in the user profile is not a good idea, i think it would be better to leave the user to decide, like putting a field in the Account Privacy Settings to choose if the Bookmarks are private or public, a button in the Bookmarks View would be a good idea as well.

Well this is it, i don't think i can answer properly the fourth question but i hope this assist you in any way and my apologies if it doesn't.

Have a nice day and sorry for bothering.

**Reply by ebwinters** [reply]
I appreciate the opinions on the design of bookmarks, but the purpose of this thread is to figure out Bluesky's process around designing these type of features.  Do we take community feedback in a forum like you just gave? Or will they provide us with designs? Those questions were just a few that could come up in a design discussion.

**Reply by SaySayTakamura** [reply]
Oh heck, i think i misunderstood your question.

But well the best place to opinate is under the issues of the social app repo and under the Discussions of this repo, you can make Feature Request or even feedback posts and see if people, either contributors or regular users, have an opinion on your work.

The discussions of this repo are normally about the AT Protocol which is the Base (or backend) for the Bluesky Network, so anything that affects this protocol can be posted both on the issues section and on the discussions section as well.

For anything related to the Bluesky Social App (the frontend of the network i assume), either web or mobile you go to the Social App Repo, there you don't have discussions but you can open an issue and see what people have to say about it.

Hope this answers better your question.
